src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/adapters/TreeListAdapter.js

```diff
@@ -0,0 +1,162 @@
+// @flow
+import {action, observable} from 'mobx';
+import {observer} from 'mobx-react';
+import React, {Fragment} from 'react';
+import Table from '../../../components/Table';
+import TreeStructureStrategy from '../structureStrategies/TreeStructureStrategy';
+import AbstractAdapter from './AbstractAdapter';
+import FullLoadingStrategy from "../loadingStrategies/FullLoadingStrategy";
+
+@observer
+export default class TreeListAdapter extends AbstractAdapter {
+    static LoadingStrategy = FullLoadingStrategy;
+
+    static StructureStrategy = TreeStructureStrategy;
+
+    static defaultProps = {
+        data: [],
+    };
+
+    @observable expandedRows: Array<string | number> = [];
+
+    getDataList() {
+        let dataList = [],
+            depth = 0;
+        dataList = this.flattenData(this.props.data, dataList, depth);
+
+        return dataList;
+    }
+
+    isExpanded(identifier: string | number) {
+        return this.expandedRows.includes(identifier);
+    }
+
+    flattenData(items: Array<Object>, dataList: Array<Object>, depth: number) {
```

Comment:
My gut feeling is that it would be easier to simply loop over the items in the `render` function instead of building another object.

Reply:
Did try that first but that didn't work as the Body.js not longer could set the callback function correctly. Try that with @trickreich at the hackday and first flatten the data did seems the only solution which did work then.

Or do you have a way where I can simple `array.push` the render elements and then render them as children from the Body.js.